### PR TITLE
codeintel: Implements path based filtering for result stream from Searcher/Zoekt

### PR DESCRIPTION
**Pull request by kritzcreek**
As part of pagination we record the file paths we've so far, so we can filter them out on the next page. That means we can no longer give Searcher/Zoekt a static count limit, as we need to stream past already seen files until we've reached the desired count. That's what this new Stream implements.

Requested review from Search Platform to maybe take a quick peek at the implementation of the Streamer to make sure I'm not doing something silly... Thanks!

## Test plan

Added a unit test, existing tests continue passing

### CODE REVIEW

**Comment by kritzcreek** [comment]
Sorry for the trickle of commits... It's actually ready for review now.

File: internal/codeintel/codenav/syntactic.go [hunk 1]
```diff
@@ -219,12 +222,60 @@ func symbolSearch(
...
+	filterFiles []string,
+	limit int,
+	limitFunc func(),
+) *CandidateStream {
+	return &CandidateStream{
+		limit: limit,
+		filterFunc: func(matches result.Matches) result.Matches {
+			return genslices.Filter(matches, func(match result.Match) bool {
+				if fileMatch, ok := match.(*result.FileMatch); ok {
+					// We only want to consider complete files for syntactic/search-based usages
```
Comment (antonsviridov-src) [inline internal/codeintel/codenav/syntactic.go:246]: Can you clarify why this is the case?
Does this mean files at the sort of end of limit window won't be considered here at all?

Comment (kritzcreek) [reply internal/codeintel/codenav/syntactic.go:246]: There are a lot of limits flying around here. I think I'm going to need to draw some boxes and arrows.

> Does this mean files at the sort of end of limit window won't be considered here at all?

Technically yes, but _that_ window is the `count: 100_000` (which is just a defensive upper limit)

File: internal/codeintel/codenav/syntactic.go [hunk 2]
```diff
@@ -219,12 +222,60 @@ func symbolSearch(
 	return symbolSearchResult{resultMap}, nil
 }
 
-func buildQuery(args searchArgs, queryType string, countLimit int) string {
+func buildQuery(args searchArgs, queryType string) string {
 	repoName := fmt.Sprintf("^%s$", args.repo)
 	wordBoundaryIdentifier := fmt.Sprintf("/\\b%s\\b/", args.identifier)
+	// NOTE: Not setting count at all will set a default limit of 10,000 according to the search-platform
```
Comment (jtibshirani) [inline internal/codeintel/codenav/syntactic.go:228]: Confirming this, and providing evidence: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/limits/limits.go?L12-14 :)

**Review by jtibshirani — COMMENTED** [review]
Apologies for the slow response here! I don't have full context on this PR, the approach looks good to me from a high level -- stream until you've reached your desired limit, then cancel.

Here's an example where we do something quite similar to implement Cody context on top of the search API: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/codycontext/job.go?L205-239. Note that 'limit' in this case is a file-based limit (not match-based), which makes things a bit simpler.

**Comment by kritzcreek** [comment]
Tabling this PR for now. I'll revisit this after my vacation with an approach that relies on retrying larger-and-larger search requests against zearcher, rather than trying to one-shot guess correctly.

**Pull request closed without merging**